File: src/salus/models/lab.py
from datetime import date, datetime, timezone
from typing import TYPE_CHECKING

from sqlmodel import Field, Relationship, SQLModel

from salus.utils import uuid7_str

if TYPE_CHECKING:
    from salus.models.user import User  # noqa: F401


class LabMarker(SQLModel, table=True):
    __tablename__ = "lab_marker"  # pyright: ignore[reportAssignmentType]

    code: str = Field(primary_key=True, foreign_key="metric_definition.code")
    category: str = Field(index=True)
    reference_low: float | None = Field(default=None)
    reference_high: float | None = Field(default=None)
    optimal_low: float | None = Field(default=None)
    optimal_high: float | None = Field(default=None)
    description: str | None = Field(default=None)


class LabPanel(SQLModel, table=True):
    __tablename__ = "lab_panel"  # pyright: ignore[reportAssignmentType]

    id: str | None = Field(default_factory=uuid7_str, primary_key=True)
    user_id: str | None = Field(default=None, foreign_key="user.id", index=True)
    collection_date: date = Field(default_factory=date.today)
    lab_name: str | None = Field(default=None)
    fasting: bool = Field(default=False)
    notes: str | None = Field(default=None)
    attachment_path: str | None = Field(default=None)
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime | None = Field(default=None)
    deleted_at: datetime | None = Field(default=None)

    results: list["LabResult"] = Relationship(back_populates="panel")


class LabResult(SQLModel, table=True):
    __tablename__ = "lab_result"  # pyright: ignore[reportAssignmentType]

    id: str | None = Field(default_factory=uuid7_str, primary_key=True)
    panel_id: str = Field(foreign_key="lab_panel.id", index=True)
    user_id: str | None = Field(default=None, foreign_key="user.id", index=True)
    metric_code: str = Field(foreign_key="metric_definition.code", index=True)
    value: float
    unit: str | None = Field(default=None)
    is_abnormal: bool = Field(default=False)
    reference_low: float | None = Field(default=None)
    reference_high: float | None = Field(default=None)
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime | None = Field(default=None)
    deleted_at: datetime | None = Field(default=None)

    panel: LabPanel = Relationship(back_populates="results")

File: frontend/src/lib/mutations/lab.ts
import { mutate } from '$lib/mutate';
import { db } from '$lib/db/database';
import { SELF_USER_ID } from '$lib/constants';
import { uuid7 } from '$lib/db/uuid';
import { nowIso, todayString } from '$lib/utils/datetime';
import { startOfLocalDayMs, userTimezone } from '$lib/utils/timezone';

const LAB_SOURCE = 'lab';

export interface LabResultInput {
  metric_code: string;
  value: number;
  unit?: string;
  is_abnormal?: boolean;
  reference_low?: number;
  reference_high?: number;
}

export interface LabPanelInput {
  collection_date?: string;
  lab_name?: string | null;
  fasting?: boolean;
  notes?: string | null;
  results: LabResultInput[];
}

function outOfRange(value: number, low: number | null, high: number | null): boolean {
  if (low != null && value < low) return true;
  if (high != null && value > high) return true;
  return false;
}

export async function createLabPanel(data: LabPanelInput) {
  const panelId = uuid7();
  const now = nowIso();
  const collectionDate = data.collection_date ?? todayString();
  const startTime = new Date(startOfLocalDayMs(collectionDate, userTimezone())).toISOString();

  const markers = await db.lab_marker.toArray();
  const markerByCode = new Map(markers.map((m) => [m.code, m]));

  const results = data.results.map((r) => {
    const marker = markerByCode.get(r.metric_code);
    const referenceLow = r.reference_low ?? marker?.reference_low ?? null;
    const referenceHigh = r.reference_high ?? marker?.reference_high ?? null;
    const value = Number(r.value);
    return {
      id: uuid7(),
      measurement_id: uuid7(),
      metric_code: r.metric_code,
      value,
      unit: r.unit ?? null,
      is_abnormal: r.is_abnormal ?? outOfRange(value, referenceLow, referenceHigh),
      reference_low: referenceLow,
      reference_high: referenceHigh
    };
  });

  return mutate({
    kind: 'command',
    command: 'create_lab_panel',
    queueable: true,
    payload: {
      id: panelId,
      collection_date: collectionDate,
      lab_name: data.lab_name ?? null,
      fasting: data.fasting ?? false,
      notes: data.notes ?? null,
      results: results.map((r) => ({
        id: r.id,
        measurement_id: r.measurement_id,
        metric_code: r.metric_code,
        value: r.value,
        unit: r.unit,
        is_abnormal: r.is_abnormal,
        reference_low: r.reference_low,
        reference_high: r.reference_high
      }))
    },
    optimisticTable: 'lab_panel',
    optimisticData: {
      id: panelId,
      user_id: SELF_USER_ID,
      collection_date: collectionDate,
      lab_name: data.lab_name ?? null,
      fasting: data.fasting ?? false,
      notes: data.notes ?? null,
      attachment_path: null,
      created_at: now,
      updated_at: null,
      deleted_at: null
    },
    optimisticRows: [
      {
        table: 'lab_result',
        rows: results.map((r) => ({
          id: r.id,
          panel_id: panelId,
          user_id: SELF_USER_ID,
          metric_code: r.metric_code,
          value: r.value,
          unit: r.unit,
          is_abnormal: r.is_abnormal,
          reference_low: r.reference_low,
          reference_high: r.reference_high,
          created_at: now,
          updated_at: null,
          deleted_at: null
        }))
      },
      {
        table: 'measurement',
        rows: results.map((r) => ({
          id: r.measurement_id,
          user_id: SELF_USER_ID,
          metric_code: r.metric_code,
          source_data_type: LAB_SOURCE,
          source: LAB_SOURCE,
          value_numeric: r.value,
          value_text: null,
          value_json: null,
          start_time: startTime,
          end_time: null,
          notes: null,
          external_id: r.id,
          created_at: now,
          updated_at: null,
          deleted_at: null
        }))
      }
    ],
    responseTable: 'lab_panel'
  });
}

export async function deleteLabPanel(panelId: string) {
  const results = await db.lab_result.where('panel_id').equals(panelId).toArray();
  const resultIds = results.map((r) => r.id);
  const measurements =
    resultIds.length > 0
      ? await db.measurement.where('external_id').anyOf(resultIds).toArray()
      : [];

  return mutate({
    kind: 'command',
    command: 'delete_lab_panel',
    queueable: true,
    payload: { id: panelId },
    optimisticTable: 'lab_panel',
    optimisticData: { id: panelId, deleted_at: nowIso() },
    optimisticDelete: [
      ...(resultIds.length > 0 ? [{ table: 'lab_result', ids: resultIds }] : []),
      ...(measurements.length > 0
        ? [{ table: 'measurement', ids: measurements.map((m) => m.id) }]
        : [])
    ]
  });
}
